README.md
---
page_type: sample
description: An Azure Maps Web SDK module that provides a window that displays a data set inside of a spyglass on the map.
languages:
- javascript
- typescript
products:
- azure
- azure-maps
---

# Azure Maps Spyglass module

An Azure Maps Web SDK module that provides a window that displays a data set inside of a spyglass on the map.

**Samples**

[Spyglass map visualization](https://azuremapscodesamples.azurewebsites.net/index.html?sample=Spyglass%20map%20visualization)
<br/>[<img src="https://github.com/Azure-Samples/AzureMapsCodeSamples/raw/master/AzureMapsCodeSamples/SiteResources/screenshots/Spyglass-map-visualization.jpg" height="200px">](https://azuremapscodesamples.azurewebsites.net/index.html?sample=Spyglass%20map%20visualization)

[Spyglass module options](https://azuremapscodesamples.azurewebsites.net/index.html?sample=Spyglass%20module%20options)
<br/>[<img src="https://github.com/Azure-Samples/AzureMapsCodeSamples/raw/master/AzureMapsCodeSamples/SiteResources/screenshots/Spyglass-module-options.jpg" height="200px">](https://azuremapscodesamples.azurewebsites.net/index.html?sample=Spyglass%20module%20options)

## Getting started

Download the project and copy the `azure-maps-spyglass` JavaScript file from the `dist` folder into your project. 

**Usage**

```HTML
<!DOCTYPE html>
<html lang="en">
<head>
    <title></title>

    <meta charset="utf-8" />
    <meta http-equiv="x-ua-compatible" content="IE=Edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />

    <!-- Add references to the Azure Maps Map control JavaScript and CSS files. -->
    <link rel="stylesheet" href="https://atlas.microsoft.com/sdk/javascript/mapcontrol/2/atlas.min.css" type="text/css" />
    <script src="https://atlas.microsoft.com/sdk/javascript/mapcontrol/2/atlas.min.js"></script>
    
    <!-- Add reference to the spyglass window module. -->
    <script src="../dist/azure-maps-spyglass.min.js"></script>

    <script type='text/javascript'>
        var primaryMap, spyMap;

        function GetMap() {
            //Add your Azure Maps key to the map SDK. Get an Azure Maps key at https://azure.com/maps. NOTE: The primary key should be used as the key.
            var authOptions = {
				authType: 'subscriptionKey',
				subscriptionKey: '<Your Azure Maps Key>'
            };

            //Initialize a map instances.
            primaryMap = new atlas.Map('primaryMap', {
                center: [-87.65, 41.87],
                zoom: 11,
                view: 'Auto',
                authOptions: authOptions
            });

            spyMap = new atlas.Map('spyMap', {
				view: 'Auto',
                authOptions: authOptions
            });

            //Create an insteance of the spyglass.
            new atlas.Spyglass(primaryMap, spyMap, {
                size: 400,
                borderColor: '#555555',
                borderWidth: 8
            });

            //Wait until the spyMap resources are ready and then load some data into is.
            spyMap.events.add('ready', function () {

                //Add some data to the map to peak into with the spy glass.

                //In this example we overlay of an old map image to the spyglass map.
                spyMap.layers.add(new atlas.layer.ImageLayer({
                    url: 'https://azuremapscodesamples.azurewebsites.net/Common/images/image-overlays/Chicago_1872_Map.png',
                    coordinates: [
                        [-87.732, 41.938], //Top Left Corner
                        [-87.592, 41.9381], //Top Right Corner
                        [-87.589, 41.811], //Bottom Right Corner
                        [-87.7298, 41.8105]  //Bottom Left Corner
                    ]
                }));
            });
        }
    </script>
    <style>
        html, body {
    		width: 100%;
    		height: 100%;
    		padding: 0;
    		margin: 0;
    	}
	
        .mapContainer {
            position: relative;
            width: 100%;
            height: 100%;
        }
    </style>
</head>
<body onload="GetMap()">
    <div class="mapContainer">
        <div id="primaryMap" class="mapContainer"></div>
        <div id="spyMap"></div>
    </div>
</body>
</html>
```

## API Reference

### Spyglass class

Namespace: `atlas`

A control that provides a window into a data set inside of a spy glass on the map.

**Contstructor**

> `Spyglass(primaryMap: azmaps.Map, spyMap: azmaps.Map, options?: SpyglassOptions)`

* primaryMap - The primary map the spy glass will be overlaid on top of.
* spyMap - The map to be used as the spy glass.
* options - The options for the control.

**Methods** 

| Name | Return type | Description |
|------|-------------|-------------|
| `dispose()` | | Dispose the control and clean up its resources. |
| `getOptions()` | `SpyglassOptions` | Gets the options of the Spyglass control. |
| `setOptions(options: SpyglassOptions)` | | Sets the options of the spyglass control. |

### SpyglassOptions interface

Options for the Spyglass.

**Properties** 

| Name | Type | Description |
|------|------|-------------|
| `borderColor` | `string` | The color of the border. Can be any CSS3 color. Default: `'#555555'` (drak grey) |
| `borderWidth` | `number` | The width of the border in pixels. Default: `5` |
| `opacity` | `number` | The opacity of the map in the spyglass. Default: `1` |
| `shape` | `'circle'` \| `'square'` | The shape of the spyglass. Can be 'circle' or 'square'. Default: `'circle'` |
| `size` | `number` | The size width/height of the shape. Default: `350` |

## Related Projects

* [Azure Maps Web SDK Open modules](https://github.com/microsoft/Maps/blob/master/AzureMaps.md#open-web-sdk-modules) - A collection of open source modules that extend the Azure Maps Web SDK.
* [Azure Maps Web SDK Samples](https://github.com/Azure-Samples/AzureMapsCodeSamples)
* [Azure Maps Gov Cloud Web SDK Samples](https://github.com/Azure-Samples/AzureMapsGovCloudCodeSamples)
* [Azure Maps & Azure Active Directory Samples](https://github.com/Azure-Samples/Azure-Maps-AzureAD-Samples)
* [List of open-source Azure Maps projects](https://github.com/microsoft/Maps/blob/master/AzureMaps.md)

## Additional Resources

* [Azure Maps (main site)](https://azure.com/maps)
* [Azure Maps Documentation](https://docs.microsoft.com/azure/azure-maps/index)
* [Azure Maps Blog](https://azure.microsoft.com/blog/topics/azure-maps/)
* [Microsoft Q&A](https://docs.microsoft.com/answers/topics/azure-maps.html)
* [Azure Maps feedback](https://feedback.azure.com/forums/909172-azure-maps)

## Contributing

We welcome contributions. Feel free to submit code samples, file issues and pull requests on the repo and we'll address them as we can. 
Learn more about how you can help on our [Contribution Rules & Guidelines](https://github.com/Azure-Samples/azure-maps-spyglass/blob/main/CONTRIBUTING.md). 

You can reach out to us anytime with questions and suggestions using our communities below:
* [Microsoft Q&A](https://docs.microsoft.com/answers/topics/azure-maps.html)
* [Azure Maps feedback](https://feedback.azure.com/forums/909172-azure-maps)

This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). 
For more information, see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or 
contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

## License

MIT
 
See [License](https://github.com/Azure-Samples/azure-maps-spyglass/blob/main/LICENSE.md) for full license text.

CHANGELOG.md
## azure-maps-spyglass Changelog

<a name="0.0.1"></a>
# 0.0.1 (2020-10-29)

**Features**

- Initial release of spyglass.



typings/index.d.ts
import * as azmaps from 'azure-maps-control';

declare namespace atlas {    

    /** Options for the Spyglass. */
    export interface SpyglassOptions {

        /** The color of the border. Can be any CSS3 color. Default: '#555555' (drak grey) */
        borderColor?: string;

        /** The width of the border in pixels. Default: 5 */
        borderWidth?: number;

        /** The opacity of the map in the spyglass. Default: 1 */
        opacity?: number;

        /** The shape of the spyglass. Can be 'circle' or 'square'. Default: 'circle' */
        shape?: 'circle' | 'square';

        /** The size width/height of the shape. Default: 350 */
        size?: number;
    }

    /** A control that provides a window into a data set inside of a spy glass on the map. */
    export class Spyglass {
        /****************************
         * Constructor
         ***************************/

        /**
            * A control that provides a window into a data set inside of a spy glass on the map.
            * @param primaryMap The primary map the spy glass will be overlaid on top of.
            * @param spyMap The map to be used as the spy glass.
            * @param options The options for the control
            */
        constructor(primaryMap: azmaps.Map, spyMap: azmaps.Map, options?: SpyglassOptions);

        /****************************
            * Public Methods
            ***************************/

        /** Dispose the control and clean up its resources. */
        public dispose(): void;

        /** Gets the options of the Spyglass control. */
        public getOptions(): SpyglassOptions;

        /**
            * Sets the options of the spyglass control.
            * @param options The options to set.
            */
        public setOptions(options: SpyglassOptions): void;
    }
}

export = atlas;

src/Spyglass.ts
import * as azmaps from "azure-maps-control";
import { SpyglassOptions } from './SpyglassOptions';

 /** A control that provides a window into a data set inside of a spy glass on the map. */
export class Spyglass {
     /****************************
     * Private Properties
     ***************************/

    private _primaryMap: azmaps.Map;
    private _spyMap: azmaps.Map;
    private _options: SpyglassOptions = {
        borderColor: '#555555',
        borderWidth: 5,
        opacity: 1,
        shape: 'circle',
        size: 350
    };
    private _syncEvents: any[] = [];
    private _maps: azmaps.Map[] = [];

    /****************************
    * Constructor
    ***************************/

    /**
     * A control that provides a window into a data set inside of a spy glass on the map.
     * @param primaryMap The primary map the spy glass will be overlaid on top of.
     * @param spyMap The map to be used as the spy glass.
     * @param options The options for the control
     */
    constructor(primaryMap: azmaps.Map, spyMap: azmaps.Map, options?: SpyglassOptions) {
        const self = this;
        self._primaryMap = primaryMap;
        self._spyMap = spyMap;

        const container = <HTMLDivElement>self._spyMap.getMapContainer();
        container.style.position = 'absolute';

        self._spyMap.events.add('ready', () => {
            (<HTMLDivElement>container.getElementsByClassName('azure-map-logo')[0]).style.display = 'none';
            (<HTMLDivElement>container.getElementsByClassName('map-copyright')[0]).style.display = 'none';
        });

        self.setOptions(self._options);

        if (options) {
            self.setOptions(options);
        }

        self._maps = [self._primaryMap, self._spyMap];

        //Bind sync events and synchronize the map views.
        self._maps.forEach((map, index) => {
            self._syncEvents[index] = self._synchronizeMaps.bind(self, map);
        });

        //Sync all map views with the first map.
        self._syncEvents[0]();

        //Attach the map move handler.
        self._attachMapMoveHandlers();
    }

    /****************************
     * Public Methods
     ***************************/

    /** Dispose the control and clean up its resources. */
    public dispose(): void {
        const self = this;
        self._detachMapMoveHandlers();

        self._options = null;

        self._spyMap = null;
        self._primaryMap = null;
    }

    /** Gets the options of the Spyglass control. */
    public getOptions(): SpyglassOptions {
        return this._options;
    }

    /**
     * Sets the options of the spyglass control.
     * @param options The options to set.
     */
    public setOptions(options: SpyglassOptions): void {
        const spyMap = this._spyMap;
        const spyMapStyle = (<HTMLDivElement>spyMap.getCanvasContainer()).style;
        const containerStyle = spyMap.getMapContainer().style;
        const opt = this._options;

        if (typeof options.shape === 'string' && (options.shape === 'circle' || options.shape === 'square')) {
            if (options.shape === 'circle') {
                containerStyle.borderRadius = '50%';
                spyMapStyle.borderRadius = '50%';
            } else {
                containerStyle.borderRadius = '0';
                spyMapStyle.borderRadius = '0';
            }

            opt.shape = options.shape;
        }

        if (typeof options.borderWidth === 'number') {
            options.borderWidth = Math.max(options.borderWidth, 0);

            containerStyle.borderStyle = 'solid';
            containerStyle.borderWidth = options.borderWidth + 'px';
            opt.borderWidth = options.borderWidth;
        }

        if (typeof options.borderColor === 'string') {
            containerStyle.borderColor = options.borderColor;
            opt.borderColor = options.borderColor;
        }
        
        if (typeof options.opacity === 'number') {
            options.opacity = Math.max(Math.min(options.opacity, 1), 0);

            spyMapStyle.opacity = options.opacity + '';
            opt.opacity = options.opacity;
        }

        if (typeof options.size === 'number') {
            options.size = Math.max(options.size, 1);

            const offset = (options.size) / 2 + opt.borderWidth;

            containerStyle.left = 'calc(50% - ' + offset + 'px)';
            containerStyle.top = 'calc(50% - ' + offset + 'px)';

            spyMap.resize(options.size, options.size);
            opt.size = options.size;
        }
    }

    /****************************
     * Private Methods
     ***************************/

    /** Attach map move handlers to the maps to synchronize them. */
    private _attachMapMoveHandlers() {
        this._maps.forEach((map, index) => {
            map.events.add('move', this._syncEvents[index]);
        });
    }

    /** Detach map move handlers to the maps. */
    private _detachMapMoveHandlers() {
        this._maps.forEach((map, index) => {
            map.events.remove('move', this._syncEvents[index]);
        });
    }

    /**
     * Synchronize all maps with a base map.
     * @param baseMap The base map to synchronize with. 
     */
    private _synchronizeMaps(baseMap: azmaps.Map) {
        const self = this;
        const targetMaps = self._maps.filter(function (m, i) { return m !== baseMap; });

        self._detachMapMoveHandlers();
        const cam = baseMap.getCamera();

        targetMaps.forEach(function (targetMap) {
            targetMap.setCamera({
                center: cam.center,
                zoom: cam.zoom,
                bearing: cam.bearing,
                pitch: cam.pitch,
                type: 'jump'
            });
        });
        self._attachMapMoveHandlers();
    }
}